File: backend/models/ServiceRequest.js
const mongoose = require('mongoose');

const ServiceRequestSchema = new mongoose.Schema({
  customerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  mechanicId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    default: null,
  },
  problemType: {
    type: String,
    enum: [
      'engineIssues',
      'brakeIssues', 
      'fuelIssues',
      'tirePuncture',
      'lockIssues',
      'electricalIssues',
      'engineLight',
      'towMe',
      'other'
    ],
    required: true,
  },
  description: {
    type: String,
    required: [true, 'Problem description is required'],
    maxlength: [500, 'Description cannot be more than 500 characters'],
  },
  vehicleNumber: {
    type: String,
    uppercase: true,
    trim: true,
  },
  customerLocation: {
    type: {
      type: String,
      enum: ['Point'],
      default: 'Point',
    },
    coordinates: {
      type: [Number], // [longitude, latitude]
      required: true,
    },
    address: {
      type: String,
      required: true,
    },
  },
  images: [{
    type: String, // URLs to uploaded images
  }],
  paymentMethod: {
    type: String,
    enum: ['cash', 'online'],
    required: true,
  },
  status: {
    type: String,
    enum: ['pending', 'accepted', 'inProgress', 'completed', 'cancelled'],
    default: 'pending',
  },
  estimatedCost: {
    type: String,
    default: null,
  },
  finalCost: {
    type: Number,
    default: null,
  },
  acceptedAt: {
    type: Date,
    default: null,
  },
  completedAt: {
    type: Date,
    default: null,
  },
  rating: {
    type: Number,
    min: 1,
    max: 5,
    default: null,
  },
  review: {
    type: String,
    maxlength: [200, 'Review cannot be more than 200 characters'],
  },
}, {
  timestamps: true,
});

// Create geospatial index for location-based queries
ServiceRequestSchema.index({ customerLocation: '2dsphere' });

// Index for efficient queries
ServiceRequestSchema.index({ customerId: 1, createdAt: -1 });
ServiceRequestSchema.index({ mechanicId: 1, createdAt: -1 });
ServiceRequestSchema.index({ status: 1, createdAt: -1 });

// Virtual for calculating elapsed time
ServiceRequestSchema.virtual('elapsedTime').get(function() {
  if (this.status === 'completed' && this.completedAt) {
    return this.completedAt - this.createdAt;
  }
  return Date.now() - this.createdAt;
});

// Method to calculate distance from a point
ServiceRequestSchema.methods.distanceFrom = function(longitude, latitude) {
  const R = 6371; // Earth's radius in kilometers
  const dLat = (latitude - this.customerLocation.coordinates[1]) * Math.PI / 180;
  const dLon = (longitude - this.customerLocation.coordinates[0]) * Math.PI / 180;
  const a = 
    Math.sin(dLat/2) * Math.sin(dLat/2) +
    Math.cos(this.customerLocation.coordinates[1] * Math.PI / 180) * 
    Math.cos(latitude * Math.PI / 180) * 
    Math.sin(dLon/2) * Math.sin(dLon/2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
  return R * c;
};

module.exports = mongoose.model('ServiceRequest', ServiceRequestSchema);

File: lib/models/service_request.dart
import 'dart:convert';

class ServiceRequest {
  final String id;
  final String customerId;
  final String? mechanicId;
  final ProblemType problemType;
  final String description;
  final String? vehicleNumber;
  final Location customerLocation;
  final List<String> images;
  final PaymentMethod paymentMethod;
  final RequestStatus status;
  final DateTime createdAt;
  final DateTime? acceptedAt;
  final DateTime? completedAt;
  final String? estimatedCost;
  final String? finalCost;

  ServiceRequest({
    required this.id,
    required this.customerId,
    this.mechanicId,
    required this.problemType,
    required this.description,
    this.vehicleNumber,
    required this.customerLocation,
    required this.images,
    required this.paymentMethod,
    required this.status,
    required this.createdAt,
    this.acceptedAt,
    this.completedAt,
    this.estimatedCost,
    this.finalCost,
  });

  factory ServiceRequest.fromJson(Map<String, dynamic> json) {
    return ServiceRequest(
      id: json['_id'] ?? '',
      customerId: json['customerId'] ?? '',
      mechanicId: json['mechanicId'],
      problemType: ProblemType.values.firstWhere(
        (e) => e.toString().split('.').last == json['problemType'],
        orElse: () => ProblemType.other,
      ),
      description: json['description'] ?? '',
      vehicleNumber: json['vehicleNumber'],
      customerLocation: Location.fromJson(json['customerLocation']),
      images: List<String>.from(json['images'] ?? []),
      paymentMethod: PaymentMethod.values.firstWhere(
        (e) => e.toString().split('.').last == json['paymentMethod'],
        orElse: () => PaymentMethod.cash,
      ),
      status: RequestStatus.values.firstWhere(
        (e) => e.toString().split('.').last == json['status'],
        orElse: () => RequestStatus.pending,
      ),
      createdAt: DateTime.parse(json['createdAt']),
      acceptedAt: json['acceptedAt'] != null ? DateTime.parse(json['acceptedAt']) : null,
      completedAt: json['completedAt'] != null ? DateTime.parse(json['completedAt']) : null,
      estimatedCost: json['estimatedCost'],
      finalCost: json['finalCost'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'customerId': customerId,
      'mechanicId': mechanicId,
      'problemType': problemType.toString().split('.').last,
      'description': description,
      'vehicleNumber': vehicleNumber,
      'customerLocation': customerLocation.toJson(),
      'images': images,
      'paymentMethod': paymentMethod.toString().split('.').last,
      'status': status.toString().split('.').last,
      'createdAt': createdAt.toIso8601String(),
      'acceptedAt': acceptedAt?.toIso8601String(),
      'completedAt': completedAt?.toIso8601String(),
      'estimatedCost': estimatedCost,
      'finalCost': finalCost,
    };
  }
}

class Location {
  final double latitude;
  final double longitude;
  final String address;

  Location({
    required this.latitude,
    required this.longitude,
    required this.address,
  });

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      latitude: json['latitude']?.toDouble() ?? 0.0,
      longitude: json['longitude']?.toDouble() ?? 0.0,
      address: json['address'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'latitude': latitude,
      'longitude': longitude,
      'address': address,
    };
  }
}

enum ProblemType {
  engineIssues,
  brakeIssues,
  fuelIssues,
  tirePuncture,
  lockIssues,
  electricalIssues,
  engineLight,
  towMe,
  other,
}

enum PaymentMethod {
  cash,
  online,
}

enum RequestStatus {
  pending,
  accepted,
  inProgress,
  completed,
  cancelled,
}